src/e3/anod/qualifiers_manager.py

```diff
@@ -1,40 +1,210 @@
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, cast
+from typing import TYPE_CHECKING
 from hashlib import sha1
-
 from e3.anod.error import AnodError
+import abc
+import re
+import logging
 
-import string
 
 if TYPE_CHECKING:
-    from typing import (
-        Literal,
-        TypedDict,
-    )
     from e3.anod.spec import Anod
 
-    # Declare types
-    TAG_QUALIFIER_TYPE = Literal["tag"]
-    KEY_VALUE_QUALIFIER_TYPE = Literal["key-value"]
 
-    class QualifierDeclaration(TypedDict):
-        # Fields common to all qualifiers
-        description: str
-        test_only: bool
-        repr_in_hash: bool
-        repr_alias: str
+VALID_NAME = re.compile(r"^[a-zA-Z0-9_.-]+$")
+
+
+def check_valid_name(name: str, value_kind: str, origin: str) -> str:
+    """Check if a value is a valid qualifier name or component name.
+
+    A valid name is any non empty string containg alphanumeric character, dot,
+    dashes and underscores.
+
+    :param name: the name to check
+    :param error_msg: an error msg
+    :return: name if name is valid otherwise raise AnodError
+    """
+    if not isinstance(name, str) or not re.search(VALID_NAME, name):
+        raise AnodError(f"{origin}: Invalid {value_kind} '{name}'")
+    return name
+
+
+class QualifierDeclaration(metaclass=abc.ABCMeta):
+    """Root class for qualifiers declaration."""
+
+    def __init__(
+        self,
+        origin: str,
+        name: str,
+        description: str,
+        repr_in_hash: bool = False,
+        repr_name: str | None = None,
+    ) -> None:
+        """Initialize a qualifier declaration.
+
+        :param origin: a string giving the origin of the declaration
+            (used in error messages)
+        :param name: qualifier name
+        :param description: help message
+        :param repr_in_hash: if True the qualifier representation is hide
+            into a common hash
+        :param repr_name: an alias for name used to compute string representation
+        """
+        self.origin = origin
+        self.name = check_valid_name(
+            name, value_kind="qualifier declaration name", origin=self.origin
+        )
+        self.description = description
+        self.repr_in_hash = repr_in_hash
+        self.repr_name = (
+            check_valid_name(
+                repr_name,
+                value_kind="qualifier declaration alias name",
+                origin=self.origin,
+            )
+            if repr_name is not None
+            else self.name
+        )
+
+    @property
+    def default(self) -> str | bool | None:
+        """Return default value for qualifier.
+
+        :return: if None is returned it means the qualifier has not
+            default Value. Otherwise the default value is returned.
+        """
+        return None  # all: no cover
+
+    @abc.abstractmethod
+    def value(self, value: str | bool) -> str | bool:
```

Comment:
This method is very strange and the naming is not very intuitive as this method seems just validate the input regarding authorized choices, so maybe a name like check_value will be more explicit and make it more understandable?

Reply:
I can change that later on as this is internal

Reply:
Ok fine by me